keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -358,6 +284,61 @@ adding functions to interfaces is a change we have made in the past, and develop
 that have alternate implementations of the interface will usually get a compiler
 error in this case, which is relatively trivial to resolve
 
+### read/modify/write loop support
+
+While it is
+[recommended](https://kubernetes.io/docs/reference/using-api/server-side-apply/)
+that apply clients use "fully specified intent" rather than
+read/modify/write loops, there are many existing controllers that use
+a get/modify/update loop today, and are not easy to convert to the
+"fully specified intend" approach.
+
+For such controllers, providing a convenient way to do a
+"get-previously-applied/modify/apply" loop is pragmatic. It allows
+these controller to benefit from apply by sending a minimal apply
+patch (as opposed to sending the entire object for update) and to
+reduce the odds of conflict with other controllers.
+
+To support this, the builders will include "BuildApply" utility functions
+function. For example:
+
+```
+fieldManger := "my-field-manager"
+// 1. read
+deployment, err := client.AppsV1().Deployments(ns).Get("deployment-name", metav1.GetOptions{})
+if err != nil {
+  // handle err
+}
+applyConfig, err := appsv1apply.BuildDeploymentApply(deployment, fieldManager)
+if err != nil {
+  // handle err
+}
+
+// 2. modify
+applyConfig.GetSpec().Replicas(10)
```

Comment:
I just realized that getters do have `Get` but setters omit the verb? I would have expected the opposite.

Reply:
The idea was to optimize for constructing apply configurations (setting fields) by omitting `Set` on functions and `New` on constructors. But maybe that goes against go conventions too much? Curious what others think.

Reply:
> The idea was to optimize for constructing apply configurations (setting fields) by omitting `Set` on functions and `New` on constructors. But maybe that goes against go conventions too much? Curious what others think.

That's kind of weird.  I think I'd keep `Set`, but I don't think it blocks the KEP overall.

Reply:
I think this line could be `applyConfig.Spec(applyv1.PodSpec().Replicas(10))` -- you can keep two layers and/or merge what is given with what you have. Then you don't have to have two ways of setting things.

I'd expect the GetFoo functions to actually return the current field value so that they can be used in if statements etc.

Reply:
Yeah, not only Go conventions but the common style in the K8s API.
The K8s interfaces have Get and Set (while something I remember about Go said you shouldn't do Get).
So I'd prefer to not mix that up too much.
On the other hand it's easy to read and makes it cleaner...

One thing that's commonly used for patterns like this in Go is `WithReplicas()` instead of Set (e.g. context.WithCancel).